### DISCUSSION TITLE
Cannot calculate  TENSOR_MECHANICS correctly

### DISCUSSION THREAD

**Discussion by ScouperLH** [idaholab/moose#18531] [Q&A Modules: General] [answered]
Hi, I am new to MOOSE. I followed the tutorial on youtube（https://www.youtube.com/watch?v=L9plLYkAbvQ&list=TLPQMDcwNzIwMjEvnbj3LWAiyQ&index=6）to do a FEM calculation of the piston, and used the correct files in the tutorial, no matter whether I updated the MOOSE and Conda in advance, my MOOSE results are as shown in the picture. Before the calculation, I have modified the Makefile, opened TENSOR_MECHANICS, and recompiled. 
I want to know how to calculate it correctly?
Thank you!
![error1](https://user-images.githubusercontent.com/35528725/128319491-80de6aed-ca18-4f93-b7e0-a105c0d6e4d9.jpg)
![error2](https://user-images.githubusercontent.com/35528725/128319503-c065cd26-48bb-448c-83c8-d03acd1732ba.jpg)
![error3](https://user-images.githubusercontent.com/35528725/128319507-e9876bd5-9e3a-47ea-9b62-5ad958582b02.jpg)

**Comment by ngrilli** [comment]
@ScouperLH 
You need to share the input file for help.
It looks like your simulation runs though.
Nicolò

**Reply by ScouperLH** [reply]
thank you for your reply! My input file is as follows. I don't see the kernel file in the app after compiling. Is it because I am using the built-in TENSOR_MECH of MOOSE?

[Mesh]
  displacements = 'disp_x disp_y disp_z' #Define displacements for deformed mesh
  type = FileMesh #Read in mesh from file
  file = piston_coarse.e
[]


[./Adaptivity]
  steps = 1
  #initial_steps = 2
  max_h_level = 3
  cycles_per_step = 1
  initial_marker = uniform
  marker = errorFraction
  [./Markers]
    [./uniform]
      type = UniformMarker
      mark = refine
    [../]
    
    [./errorFraction]
      type = ErrorFractionMarker
      coarsen = 0.5
      indicator = gradientJump
      refine = 0.5
    [] 
 [../]

  [./Indicators]
    [./gradientJump]
      type = GradientJumpIndicator
      variable = disp_y
    []
  []
[../]

[Variables]
  [./disp_x]
    order = FIRST
    family = LAGRANGE
  [../]
  [./disp_y]
    order = FIRST
    family = LAGRANGE
  [../]
  [./disp_z]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[Kernels]
  [./TensorMechanics]
    #Stress divergence kernels
    displacements = 'disp_x disp_y disp_z'
  [../]
[]

[AuxVariables]
  [./von_mises]
    #Dependent variable used to visualize the Von Mises stress
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[AuxKernels]
  [./von_mises_kernel]
    #Calculates the von mises stress and assigns it to von_mises
    type = RankTwoScalarAux
    variable = von_mises
    rank_two_tensor = stress
    execute_on = timestep_end
    scalar_type = VonMisesStress
  [../]
[]

[Functions]
  [./rampLinear]
    type = ParsedFunction
    value = t*550e5
  []
[]

[BCs]
  [./Pressure]
    [./load]
      #Applies the pressure
      boundary = load_surf
      function = rampLinear
      disp_x = disp_x
      disp_y = disp_y
      disp_z = disp_z
    [../]
  [../]
  [./symmetry_x]
    #Applies symmetry on the xmin faces
    type = DirichletBC
    variable = disp_x
    boundary = 'xmin'
    value = 0.0
  [../]
  [./hold_y]
    #Anchors the bottom against deformation in the y-direction
    type = DirichletBC
    variable = disp_y
    boundary = 'ymin'
    value = 0.0
  [../]
  [./symmetry_z]
    #Applies symmetry on the zmin faces
    type = DirichletBC
    variable = disp_z
    boundary = 'zmin'
    value = 0.0
  [../]
[]

[Materials]
  active = 'density_steel stress strain elasticity_tensor_steel'
  [./elasticity_tensor_steel]
    #Creates the elasticity tensor using steel parameters
    youngs_modulus = 210e9 #Pa
    poissons_ratio = 0.3
    type = ComputeIsotropicElasticityTensor
    block = 1
  [../]
  [./strain]
    #Computes the strain, assuming small strains
    type = ComputeFiniteStrain
    block = 1
    displacements = 'disp_x disp_y disp_z'
  [../]
  [./stress]
    #Computes the stress, using linear elasticity
    type = ComputeFiniteStrainElasticStress
    block = 1
  [../]
  [./density_steel]
    #Defines the density of steel
    type = GenericConstantMaterial
    block = 1
    prop_names = density
    prop_values = 7850 # kg/m^3
  [../]
[]

[Preconditioning]
  [./SMP]
    #Creates the entire Jacobian, for the Newton solve
    type = SMP
    full = true
  [../]
[]

[Executioner]
  #We solve a steady state problem using Newton's iteration
  type = Transient
  solve_type = NEWTON
  nl_rel_tol = 1e-9
  l_max_its = 30
  l_tol = 1e-4
  nl_max_its = 10
  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart'
  petsc_options_value = 'hypre boomeramg 31'
  dt = 0.1
  num_steps = 10
[]

[Outputs]
  exodus = true
  perf_graph = true
[]

**Comment by ajacquey** [comment, marked as answer]
You should modify your pressure BC. 

Looking at the source code of the [PressureAction](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/actions/PressureAction.C), you should provide the displacements as a vector `displacements = 'disp_x disp_y disp_z'` instead of providing the three displacement variables separately.

Your BC should look like that:

```
[./Pressure]
  [./load]
    #Applies the pressure
	boundary = load_surf
	function = rampLinear
	displacements = 'disp_x disp_y disp_z'
  [../]
```

**Reply by ScouperLH** [reply]
Thank you! I modified it accordingly, and it works!

**Comment by hugary1995** [comment]
@GregVernon this is a nice tutorial. I suggest to check it into the tensor mechanics module as an example, and enable at least syntax check so that it doesn't break in the future.

**Reply by GregVernon** [reply]
Will do!

**Reply by hugary1995** [reply]
Now that this is merged, I think it'd be great if you can somehow point viewers of that YouTube tutorial to the updated input file. I'm not a YouTube expert, but I think that's doable. Anyways, it's your call.

**Reply by GregVernon** [reply]
Yep - I'll do that today!
